api/all/src/main/java/io/opentelemetry/api/GlobalOpenTelemetry.java

```diff
@@ -67,14 +68,31 @@ public static OpenTelemetry get() {
    * Sets the {@link OpenTelemetry} that should be the global instance. Future calls to {@link
    * #get()} will return the provided {@link OpenTelemetry} instance. This should be called once as
    * early as possible in your application initialization logic, often in a {@code static} block in
-   * your main class.
+   * your main class. It should only be called once - an attempt to call it a second time will
+   * result in an error. If trying to set the global {@link OpenTelemetry} multiple times in tests,
+   * use {@link #resetForTest()} between them.
+   *
+   * <p>If you are using the OpenTelemetry SDK, you should generally use {@code
+   * OpenTelemetrySdk.builder().buildAndRegisterGlobal()} instead of calling this method directly.
    */
   public static void set(OpenTelemetry openTelemetry) {
-    globalOpenTelemetry = openTelemetry;
+    synchronized (mutex) {
+      if (globalOpenTelemetry != null) {
+        throw new IllegalStateException(
+            "GlobalOpenTelemetry.set has already been called, previous invocation set to cause of "
+                + "this exception.",
+            setGlobalCaller);
+      }
+      globalOpenTelemetry = openTelemetry;
+      setGlobalCaller = new Throwable();
+    }
   }
 
-  // for testing
-  static void reset() {
+  /**
+   * Unsets the global {@link OpenTelemetry}. This is only meant to be used from tests which need to
+   * reconfigure {@link OpenTelemetry}.
+   */
+  public static void resetForTest() {
```

Comment:
oof. this being public makes me pretty uncomfortable. Is there any way we can restrict this to being package access again? Perhaps with a same-package testing bridge?

Reply:
I moved to a separate class in the same package, not sure how much better it is, maybe a little.